src/Microsoft.OData.Edm/Vocabularies/Annotations/EdmTermExtensions.cs

```diff
@@ -0,0 +1,446 @@
+                    {
+                        return new EdmGuidConstant(gd.Value);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                case EdmPrimitiveTypeKind.Int16:
+                case EdmPrimitiveTypeKind.Int32:
+                case EdmPrimitiveTypeKind.Int64:
+                    if (EdmValueParser.TryParseInt(defaultValue, out int? intNum))
+                    {
+                        return new EdmIntegerConstant(intNum.Value);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                case EdmPrimitiveTypeKind.Date:
+                    if (EdmValueParser.TryParseDate(defaultValue, out Date? dt))
+                    {
+                        return new EdmDateConstant(dt.Value);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                case EdmPrimitiveTypeKind.Byte:
+                case EdmPrimitiveTypeKind.SByte:
+                case EdmPrimitiveTypeKind.Stream:
+                case EdmPrimitiveTypeKind.Geography:
+                case EdmPrimitiveTypeKind.GeographyPoint:
+                case EdmPrimitiveTypeKind.GeographyLineString:
+                case EdmPrimitiveTypeKind.GeographyPolygon:
+                case EdmPrimitiveTypeKind.GeographyCollection:
+                case EdmPrimitiveTypeKind.GeographyMultiPolygon:
+                case EdmPrimitiveTypeKind.GeographyMultiLineString:
+                case EdmPrimitiveTypeKind.GeographyMultiPoint:
+                case EdmPrimitiveTypeKind.Geometry:
+                case EdmPrimitiveTypeKind.GeometryPoint:
+                case EdmPrimitiveTypeKind.GeometryLineString:
+                case EdmPrimitiveTypeKind.GeometryPolygon:
+                case EdmPrimitiveTypeKind.GeometryCollection:
+                case EdmPrimitiveTypeKind.GeometryMultiPolygon:
+                case EdmPrimitiveTypeKind.GeometryMultiLineString:
+                case EdmPrimitiveTypeKind.GeometryMultiPoint:
+                case EdmPrimitiveTypeKind.PrimitiveType:
+                case EdmPrimitiveTypeKind.None:
+                default:
+                    break;
+            }
+            throw Error.NotSupported();
+        }
+
+        /// <summary>
+        /// Parses a <paramref name="defaultValue"/> into an IEdmExpression value of the correct CSDL type.
+        /// </summary>
+        /// <param name=“typeReference”>The type of value.</param>
+        /// <param name=“defaultValue”>Original default value.</param>
+        /// <returns>An IEdmExpression of type <paramref name="typeReference".</paramref></returns>
+        public static CsdlExpressionBase BuildDefaultCsdlExpression(IEdmTypeReference typeReference, string defaultValue, CsdlLocation location)

```

Comment:
>BuildDefaultCsdlExpression [](http://example.com/codeflow?start=41&length=26)

We should consider whether we can consolidate these code paths; maybe in ODL 8.0. 

Do EdmVocabularyAnnotation and CsdlVocabularyAnnotation really care that their expressions are EdmExpression/CsdlExpression rather than IEdmExpression?

Reply:
Based on a discussion I had with Sam, the distinction is important internally. CsdlExpressionBase is an intermediate wrapper -- the final wrapper is the CsdlSemantics*Expression, which implements IEdmExpression -- but it sounds like it's still needed. However, to simplify things for clients and as encouraged by Sam, I've now created a different public method that does the more external work of returning an IEdmExpression (still creating a CsdlExpressionBase internally as above).